### DISCUSSION TITLE
Is there a macro to enable [[5]] syntax for tensors?

### DISCUSSION THREAD

**Discussion by veqqq** [themetaschemer/malt#12] [Q&A]
`(tensor (tensor 5))` -> `'[[5]]` or `[[5]]` or something would be nice for interoperability.

I haven't been able to make it work and suspect it isn't possible or might require a lot of use of | in some pattern I haven't found yet.

**Comment by themetaschemer** [comment]
Hi veqqq, interesting question. A couple of remarks about it: 

1. [ ]'s are used in Racket as an alternative to ( )'s and is often used as separation for let bindings, cond clauses and the like. This was one of the reasons we only used it as notation in the book and use the preferred translation as (tensor ...). A close enough syntax that still allows you to do something similar would be to use #[ ...], which is used for vectors by default, but you could override it for tensors. 

2. In either case, you should look at Racket's reader extensions: https://docs.racket-lang.org/guide/hash-reader.html which will help you achieve the goals. 
 
Please do post whatever solution you come up with over here!

**Comment by Solaxun** [comment]
As a compromise to reader macros would something like the following helper work? You could make it a macro to avoid runtime  conversion from list->tensor if that's important.

```
(define (array->tensor a)
    (if (scalar? a) 
    (tensor a)
    (apply tensor (map array->tensor a))))

(array->tensor '((1 2 3) (4 5 6)))
```
Haven't tested it as I'm away from my computer that has malt installed.
